tests/test_readme_util.py

```diff
@@ -0,0 +1,242 @@
+### Dataset Summary
+Some text here.
+### Supported Tasks and Leaderboards
+### Languages
+"""
+
+README_MISSING_TEXT = """\
+---
+languages:
+- zh
+- en
+---
+
+# Dataset Card for My Dataset
+## Table of Contents
+Some text here.
+## Dataset Description
+Some text here.
+### Dataset Summary
+### Supported Tasks and Leaderboards
+### Languages
+"""
+
+README_MISSING_SUBSECTION = """\
+---
+languages:
+- zh
+- en
+---
+
+# Dataset Card for My Dataset
+## Table of Contents
+Some text here.
+## Dataset Description
+Some text here.
+### Dataset Summary
+Some text here.
+### Languages
+"""
+
+README_MISSING_FIRST_LEVEL = """\
+---
+languages:
+- zh
+- en
+---
+
+## Table of Contents
+Some text here.
+## Dataset Description
+Some text here.
+### Dataset Summary
+Some text here.
+### Supported Tasks and Leaderboards
+### Languages
+"""
+
+
+README_MULTIPLE_WRONG_FIRST_LEVEL = """\
+---
+languages:
+- zh
+- en
+---
+
+# Dataset Card for My Dataset
+## Table of Contents
+Some text here.
+## Dataset Description
+Some text here.
+### Dataset Summary
+Some text here.
+### Supported Tasks and Leaderboards
+### Languages
+# Dataset Card My Dataset
+"""
+
+README_WRONG_FIRST_LEVEL = """\
+---
+languages:
+- zh
+- en
+---
+
+# Dataset Card My Dataset
+## Table of Contents
+Some text here.
+## Dataset Description
+Some text here.
+### Dataset Summary
+Some text here.
+### Supported Tasks and Leaderboards
+### Languages
+"""
+
+README_EMPTY = ""
+
+README_MULTIPLE_SAME_HEADING_1 = """\
+---
+languages:
+- zh
+- en
+---
+
+# Dataset Card for My Dataset
+# Dataset Card for My Dataset
+## Table of Contents
+Some text here.
+## Dataset Description
+Some text here.
+### Dataset Summary
+Some text here.
+### Supported Tasks and Leaderboards
+### Languages
+"""
+
+
+class TestReadMeUtils(unittest.TestCase):
+    def test_from_string(self):
+        ReadMe.from_string(README_CORRECT, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_EMPTY_YAML, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_INCORRECT_YAML, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_NO_YAML, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_MISSING_TEXT, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_MISSING_SUBSECTION, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_MISSING_FIRST_LEVEL, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_MULTIPLE_WRONG_FIRST_LEVEL, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_WRONG_FIRST_LEVEL, EXPECTED_STRUCTURE)
+        with self.assertRaises(ValueError):
+            ReadMe.from_string(README_EMPTY, EXPECTED_STRUCTURE)
```

Comment:
Here you could use pytest to check for the error messages.
You can find some documentation here:
https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions

Note that pytest doesn't use the `unittest.TestCase` class. Instead you have to define a test function.
For example
```python
def test_from_string(self):
    ReadMe.from_string(README_CORRECT, EXPECTED_STRUCTURE)
    with pytest.raises(ValueError) as excinfo:
        ReadMe.from_string(README_EMPTY_YAML, EXPECTED_STRUCTURE)
    assert "empty" in excinfo
```

Does that sound good for you ?

Reply:
Also you can use `@pytest.mark.parametrize(...)` to run your test functions on all the dummy yaml you defined if it sounds more convenient for you

Reply:
Oh, I thought I was restricted to `unittest`. Cool, I'll write `pytest` test cases and also check the error messages. I assume that is better?

Reply:
That would be ideal, thanks !